sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -41,19 +87,43 @@ cdef inline INDEX_T _deg3_column(INDEX_T d, INDEX_T i, INDEX_T j, INDEX_T k,
         return ((3 * d**2 * i - 3 * d * i**2 + i ** 3 - i
                  - 3 * j**2 - 3 * j) / 6
                 + d * j + k)
+def py_deg3_column(d, i, j, k, interaction_only):
+    """Compute the index of the column for a degree 3 expansion.
 
+    d is the dimensionality of the input data, i, j and k are the indices
+    for the columns involved in the expansion.
 
-def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
-                              cnp.ndarray[INDEX_T, ndim=1] indices,
-                              cnp.ndarray[INDEX_T, ndim=1] indptr,
-                              INDEX_T d, INDEX_T interaction_only,
-                              INDEX_T degree):
+    This function is defined in Python to use `PyLong`'s arbitrary
+    precision, and is called in Cython when C types' representability
+    range is too small.
+    """
+    if interaction_only:
+        return ((3 * d**2 * i - 3 * d * i**2 + i**3
+                 + 11 * i - 3 * j**2 - 9 * j) // 6
+                + i**2 - 2 * d * i + d * j - d + k)
+    else:
+        return ((3 * d**2 * i - 3 * d * i**2 + i ** 3 - i
+                 - 3 * j**2 - 3 * j) // 6
+                + d * j + k)
+
+# TODO: `const`-qualify data, indices and indptr when Cython>=3.0 is released.
+def _csr_polynomial_expansion(
+    DATA_t[:] data,                 # IN READ-ONLY
+    INDEX_A_t[:] indices,           # IN READ-ONLY
+    INDEX_A_t[:] indptr,            # IN READ-ONLY
+    INDEX_A_t d,
+    DATA_t[:] result_data,          # OUT
+    INDEX_B_t[:] result_indices,    # OUT
+    INDEX_B_t[:] result_indptr,     # OUT
+    FLAG_t interaction_only,
+    FLAG_t degree
+):
     """
     Perform a second-degree polynomial or interaction expansion on a scipy
```

Comment:
Pre-existing (from me), but this function can do either second or third degree expansions, not just second.